### ISSUE TITLE
Re-queries on hydration after ssr using renderToPipeableStream

### ISSUE THREAD

**Issue by iamafansev** [apollographql/apollo-client#11319]
### Issue Description

version: 3.8.*
To begin with, it is better to consider the problem using an example:
**Example with useSuspenseQuery & renderToPipeableStream**
https://github.com/iamafansev/test-suspense-query

In this example, the **fetchPolicy** is set to **cache-and-network**.
We expect that there will be no duplicate requests during hydration since we have a valid cache.

However, we see repeated outgoing requests in the developer tools.

At the same time, if you set the caching policy to `cache-first`, then repeated requests are not executed. This confirms that the cache contains correct data and the request need not be executed.

When I encountered this problem, I thought that the problem was only when using new features such as `useSuspenseQuery`, but I have another example:
Example with `useQuery / getDataFromTree & renderToPipeableStream ` (workaround to use React.(Suspense / lazy))
https://github.com/iamafansev/use-query-with-lazy-ssr

There is some workaround here to use pause rendering just to wait for lazy components. In the server-side implementation, we first traverse the tree using `getDataFromTree` and only after that, we pass the tree to `renderToPipeableStream`.

Similar to the first example, a repeated request is performed here with the cache-and-network caching policy and is not performed with cache-first.
The problem also appears in version 3.7.17

### Link to Reproduction

https://github.com/iamafansev/test-suspense-query

### Reproduction Steps

- yarn install
- yarn start
- go to http://localhost:4000/
- devtools
- network tab
- We make sure there are repeat requests

**Comment by phryneas** [comment]
Hi there, lots of things to unpack here!

Unfortunately, you've fallen into a big pitfall the React team hasn't really documented well: there is currently no data fetching story for `renderToPipeableStream` in React.  
[See twitter discussion here](https://twitter.com/dan_abramov/status/1718721732247228628) (I'm really sorry that I have to give twitter as a reference here - I'd love to link you some docs, but this just hasn't been documented at all 🤦).

You can do the workaround you are currently doing - running `getDataFromTree`, gather all the data you need and then run `renderToPipeableStream` - but unless I miss some nuance here, there isn't really anything you gain from it: you essentially fully block the render, and then start a stream that could at that point in time just as well be a synchronous render (after all, you did all the async work before!).  
It's also not really something we support, since - see above - it doesn't seem like it really makes sense to use `renderToPipeableStream` under these circumstances.

I have recently talked with Josh Story about this, and there is a good chance that React 19 might include a few helpers that would be necessary for us to get this working - but as of right now, there is just no React API for us to inject into the stream, so we cannot support this.  
We do work around that and do have support for `renderToPipeableStream` in our [Next.js helper library](https://github.com/apollographql/apollo-client-nextjs), but that relies on Next.js internals that kinda make up for the lack of React features here.

TLDR: as of right now, we cannot support `renderToPipeableStream` as React is lacking the primitives for us (or any other data fetching library) to do so. 
I'm really sorry.

**Comment by kcoet** [comment]
I was stuck in the same problem for several days.

I hope the Apollo team can provide examples of suggested SSRs now that as with libraries in general, they have an examples directory.

@phryneas Do you have any suggestions what is the best current method for building SSR with Apollo? is it still renderToStaticMarkup like before?

**Comment by alvaro-cuesta** [comment]
For anyone interested, I hacked around this by only sending the `renderToPipeableStream` response on `onAllReady`.

So basically... by **not** streaming, but works for my use case since I'm just migrating from React 17 + Loadable Components and I needed `Reac.lazy` SSR support, which only `renderToPipeableStream` supports... and as a side effect the cache for suspended queries is fully populated on hydration, so there are no additional queries for me.

Sharing this in case someone is in a migration like me and got scared by this thread... there's hope!

Can't wait to have proper streaming implemented.

**Comment by moishinetzer** [comment]
I'd love to see an update on this. 

The documentation is very misleading, as React says very explicitly that `renderToString` does not support `Suspense`, so it feels hacky that we are allowing suspense boundaries while not using the pipe-able API.

I think it would be very helpful to have this ironed out in the docs.

Thanks for your hard work

**Comment by moishinetzer** [comment]
Thanks for the swift response! 

I'll give it a shot and update here on the thread. 

My context is that we have a really large production react app and we're trying to migrate from Webpack with loadable (to split up our routes), to Vite with lazy loading. 

However, lazy loading (as the user above mentioned) is only supported with the streaming APIs

**Comment by phryneas** [comment]
@moishinetzer then you probably have to go with the setup in the documentation that I linked to.

That said, I remember setting up streaming SSR with Vite to be quite the beast. You might want to look into something like Vike - they also have an [Apollo Client extension](https://github.com/vikejs/vike-react/tree/main/packages/vike-react-apollo#readme) based on our streaming package.

**Comment by moishinetzer** [comment]
I see, so for now I should be able to go ahead with my method above so long as I change all my queries to use the `useSuspenseQuery` hook instead of useQuery!

It seems to have done the trick!
